// @remotion/studio/dist/components/Timeline/sequence-props-subscription-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.acquireSequencePropsSubscription = void 0;
const remotion_1 = require("remotion");
const call_api_1 = require("../call-api");
const INITIAL_SNAPSHOT = {
    nodePath: null,
    jsxInMapCallback: false,
    props: null,
};
const entries = new Map();
let globalUnsubscribe = null;
const makeKey = (fileName, line, column) => `${fileName}|${line}|${column}`;
const nodePathsEqual = (a, b) => {
    if (!a || !b) {
        return false;
    }
    return JSON.stringify(a) === JSON.stringify(b);
};
const notify = (entry) => {
    for (const listener of entry.listeners) {
        listener(entry.snapshot);
    }
};
const handleEvent = (event) => {
    if (event.type !== 'sequence-props-updated') {
        return;
    }
    for (const entry of entries.values()) {
        if (entry.fileName !== event.fileName) {
            continue;
        }
        if (!nodePathsEqual(entry.snapshot.nodePath, event.nodePath)) {
            continue;
        }
        if (event.result.canUpdate) {
            entry.snapshot = {
                nodePath: entry.snapshot.nodePath,
                jsxInMapCallback: event.result.jsxInMapCallback,
                props: event.result.props,
            };
        }
        else {
            entry.snapshot = {
                nodePath: entry.snapshot.nodePath,
                jsxInMapCallback: false,
                props: null,
            };
        }
        notify(entry);
    }
};
const ensureGlobalListener = (subscribeToEvent) => {
    if (globalUnsubscribe) {
        return;
    }
    globalUnsubscribe = subscribeToEvent('sequence-props-updated', handleEvent);
};
const teardownGlobalListenerIfEmpty = () => {
    if (entries.size > 0) {
        return;
    }
    if (globalUnsubscribe) {
        globalUnsubscribe();
        globalUnsubscribe = null;
    }
};
const fireUnsubscribe = (fileName, nodePath, clientId) => {
    (0, call_api_1.callApi)('/api/unsubscribe-from-sequence-props', {
        fileName,
        nodePath,
        clientId,
    }).catch(() => {
        // Ignore unsubscribe errors
    });
};
const acquireSequencePropsSubscription = ({ clientId, fileName, line, column, schema, subscribeToEvent, onChange, }) => {
    const key = makeKey(fileName, line, column);
    let entry = entries.get(key);
    if (!entry) {
        const newEntry = {
            clientId,
            fileName,
            refCount: 0,
            snapshot: INITIAL_SNAPSHOT,
            listeners: new Set(),
            tornDown: false,
        };
        entries.set(key, newEntry);
        ensureGlobalListener(subscribeToEvent);
        (0, call_api_1.callApi)('/api/subscribe-to-sequence-props', {
            fileName,
            line,
            column,
            schema,
            clientId,
        })
            .then((result) => {
            if (newEntry.tornDown) {
                if (result.canUpdate) {
                    fireUnsubscribe(fileName, result.nodePath, clientId);
                }
                return;
            }
            if (result.canUpdate) {
                newEntry.snapshot = {
                    nodePath: result.nodePath,
                    jsxInMapCallback: result.jsxInMapCallback,
                    props: result.props,
                };
            }
            else {
                newEntry.snapshot = INITIAL_SNAPSHOT;
            }
            notify(newEntry);
        })
            .catch((err) => {
            if (newEntry.tornDown) {
                return;
            }
            remotion_1.Internals.Log.error(err);
            newEntry.snapshot = INITIAL_SNAPSHOT;
            notify(newEntry);
        });
        entry = newEntry;
    }
    entry.refCount += 1;
    entry.listeners.add(onChange);
    onChange(entry.snapshot);
    let released = false;
    return () => {
        if (released) {
            return;
        }
        released = true;
        const e = entry;
        e.listeners.delete(onChange);
        e.refCount -= 1;
        if (e.refCount > 0) {
            return;
        }
        e.tornDown = true;
        const resolvedNodePath = e.snapshot.nodePath;
        entries.delete(key);
        if (resolvedNodePath) {
            fireUnsubscribe(e.fileName, resolvedNodePath, e.clientId);
        }
        teardownGlobalListenerIfEmpty();
    };
};
exports.acquireSequencePropsSubscription = acquireSequencePropsSubscription;
