// @remotion/studio/dist/components/Timeline/TimelineEnumField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEnumField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const selectStyle = {
    marginLeft: 8,
    background: 'transparent',
    color: 'white',
    border: '1px solid rgba(255, 255, 255, 0.2)',
    borderRadius: 3,
    fontSize: 12,
    padding: '1px 4px',
};
const TimelineEnumField = ({ field, codeValue, effectiveValue, canUpdate, onSave, onDragValueChange, onDragEnd, }) => {
    const { fieldSchema } = field;
    if (fieldSchema.type !== 'enum') {
        throw new Error('TimelineEnumField rendered for non-enum field');
    }
    const variantKeys = Object.keys(fieldSchema.variants);
    const current = String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : fieldSchema.default);
    const onChange = (0, react_1.useCallback)((e) => {
        const newValue = e.target.value;
        if (!canUpdate || newValue === codeValue) {
            return;
        }
        onDragValueChange(field.key, newValue);
        onSave(field.key, newValue).finally(() => {
            onDragEnd();
        });
    }, [canUpdate, codeValue, field.key, onSave, onDragValueChange, onDragEnd]);
    return (jsx_runtime_1.jsx("select", { disabled: !canUpdate, value: current, onChange: onChange, style: selectStyle, children: variantKeys.map((key) => (jsx_runtime_1.jsx("option", { value: key, children: key }, key))) }));
};
exports.TimelineEnumField = TimelineEnumField;
