// @remotion/studio/dist/components/Timeline/use-sequence-props-subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSequencePropsSubscription = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const sequence_props_subscription_store_1 = require("./sequence-props-subscription-store");
const EMPTY_STATE = {
    nodePath: null,
    jsxInMapCallback: false,
};
const useSequencePropsSubscription = (sequence, originalLocation, visualModeEnabled) => {
    var _a, _b;
    var _c, _d, _e, _f;
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeOverridesContext);
    const overrideId = (_c = (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.overrideId) !== null && _c !== void 0 ? _c : null;
    const setPropStatusesForSequence = (0, react_1.useCallback)((statuses) => {
        if (!overrideId) {
            return;
        }
        setCodeValues(overrideId, statuses);
    }, [overrideId, setCodeValues]);
    const { previewServerState: state, subscribeToEvent } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const clientId = state.type === 'connected' ? state.clientId : undefined;
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const [subscriptionState, setSubscriptionState] = (0, react_1.useState)(EMPTY_STATE);
    const isMountedRef = (0, react_1.useRef)(true);
    (0, react_1.useEffect)(() => {
        isMountedRef.current = true;
        return () => {
            isMountedRef.current = false;
        };
    }, []);
    const schema = (_b = sequence.controls) === null || _b === void 0 ? void 0 : _b.schema;
    const locationSource = (_d = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source) !== null && _d !== void 0 ? _d : null;
    const locationLine = (_e = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.line) !== null && _e !== void 0 ? _e : null;
    const locationColumn = (_f = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.column) !== null && _f !== void 0 ? _f : null;
    (0, react_1.useEffect)(() => {
        if (!visualModeEnabled) {
            setPropStatusesForSequence(null);
            setSubscriptionState(EMPTY_STATE);
            return;
        }
        if (!clientId ||
            !locationSource ||
            !locationLine ||
            locationColumn === null ||
            !schema) {
            setPropStatusesForSequence(null);
            setSubscriptionState(EMPTY_STATE);
            return;
        }
        const onChange = (snapshot) => {
            setSubscriptionState({
                nodePath: snapshot.nodePath,
                jsxInMapCallback: snapshot.jsxInMapCallback,
            });
            setPropStatusesForSequence(snapshot.props);
        };
        const release = (0, sequence_props_subscription_store_1.acquireSequencePropsSubscription)({
            clientId,
            fileName: locationSource,
            line: locationLine,
            column: locationColumn,
            schema,
            subscribeToEvent,
            onChange,
        });
        return () => {
            release();
            // Only clear props on true unmount, not on re-subscribe due to
            // location changes — avoids flicker while re-subscribing.
            if (!isMountedRef.current) {
                setPropStatusesForSequence(null);
            }
        };
    }, [
        clientId,
        locationColumn,
        locationLine,
        locationSource,
        schema,
        setPropStatusesForSequence,
        subscribeToEvent,
        visualModeEnabled,
    ]);
    return subscriptionState;
};
exports.useSequencePropsSubscription = useSequencePropsSubscription;
